### PR TITLE
SEQNG-818 Fixed GNIRS bias check

### PR DESCRIPTION
**Pull request by jluhrs**
There was an issue with the check to decide if the GNIRS bias had to be reapplied. The value read from the instrument has the opposite sign.
I got a bit sidetracked. I moved the GNIRS function to check parameters before applying them, to `EpicsUtil`, and changed the GMOS code to use it. That is in the first commit.

### CODE REVIEW

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosController.scala [hunk 1]
```diff
@@ -56,7 +56,8 @@ object GmosController {
     @SuppressWarnings(Array("org.wartremover.warts.LeakingSealed"))
     case class BuiltInFPU(fpu: T#FPU) extends GmosFPU
 
-    case class GmosDisperser(disperser: T#Disperser, order: Option[DisperserOrder], lambda: Option[Length])
+    case class GmosDisperser(disperser: T#Disperser, order: Option[DisperserOrder],
+      lambda: Option[Length])
 
     case class CCConfig(
```
Comment (sraaphorst) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosController.scala]: It doesn't look like we're extending any of these, so why not make them `final case class`?

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala [hunk 2]
```diff
@@ -69,78 +61,117 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
...
         roiParameters(binning, i + 1, ROIValues.fromOCS(roi))
-      }.sequence.flatMap(_ => SeqAction.void)
+      }
+    }
+
+  private def roiParameters(binning: CCDBinning, index: Int, roi: Option[ROIValues])
+  : List[SeqAction[Unit]] = (roi, DC.rois.get(index)).mapN { (roi, r) =>
+    r.setCcdXstart1(roi.xStart.value) *>
+    r.setCcdXsize1(roi.xSize.value / binning.x.getValue) *>
+    r.setCcdYstart1(roi.yStart.value) *>
```
Comment (sraaphorst) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala]: I've never known what `*>` was used for before. This was very helpful.

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala [hunk 3]
```diff
@@ -234,7 +235,8 @@ object GnirsControllerEpics extends GnirsController {
     val coaddsWriter = smartSetParam(config.coadds, epicsSys.numCoadds,
       dcCmd.setCoadds(config.coadds))
 
-    val biasWriter =smartSetDoubleParam(biasTolerance)(encode(config.wellDepth), epicsSys.detBias,
+    // Value read from the instrument is the negative of what was set
+    val biasWriter = smartSetDoubleParam(biasTolerance)(-encode(config.wellDepth), epicsSys.detBias,
```
Comment (sraaphorst) [inline modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: Are we sure this should change from `encode` to `-encode`?

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala (hunk 2 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala]: It's a `flatMap` that ignores the input.

File: modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala (hunk 3 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala]: The arguments are: value too set, function to read back the applied value, function to apply the new value. The issue was that the value read back  from the instrument was the negative of the value set. Alternatively, I could have changed the sign of the second argument.

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosController.scala (hunk 1 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosController.scala]: Making them `final` produces a compile error: "The outer reference in this type test cannot be checked at run time."

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala (hunk 2 above)
Comment (sraaphorst) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala]: Thanks for the explanation. That will be useful in the future for me, I'm sure. Looks good!

**Pull request merged**